<launch>

	<arg name="uavname" default="fpv" />

    <arg name="imu_topic" default="/djiros/imu"/>

    <arg name="topic_odom_uav" default="/odom/uav"/>
    <arg name="topic_odom_cone" default="/odom/cone"/>

    <arg name="topic_rocking_cmd" value="/rnw/rocking_cmd"/>
    <arg name="topic_cone_state" value="/rnw/cone_state"/>
    <arg name="topic_poly_traj" value="/rnw/poly_traj" />

    <arg name="topic_trigger_insert" value="/rnw/trigger/insert" />
    <arg name="topic_trigger_zigzag" value="/rnw/trigger/zigzag" />
    <arg name="topic_trigger_topple" value="/rnw/trigger/topple" />
    <arg name="topic_trigger_rnw" value="/rnw/trigger/rnw" />
    <arg name="topic_trigger_push_init" value="/rnw/trigger/push_init" />
    <arg name="topic_trigger_adjust_nutation" value="/rnw/trigger/adjust_nutation" />
    <arg name="topic_trigger_adjust_grip" value="/rnw/trigger/adjust_grip" />
    <arg name="topic_trigger_adjust_yaw" value="/rnw/trigger/adjust_yaw" />

    <rosparam command="load" file="$(find rnw_ros)/config/rnw_config.yaml"/>

    <!-- Using dynamic_reconfigure will cause timeout -->
    <!-- Increase from 0.1 default to 0.5 to avoid entering atti mode -->
    <param name="ctrl_cmd_stream_timeout" value="0.5"/>

    <node pkg="djiros" type="djiros_node" name="djiros" output="screen" launch-prefix="" required="true">
        <param name="log_level" type="string" value="info"/>
        <param name="serial_name" type="string" value="/dev/tty.dji.n3"/>
        <param name="baud_rate" type="int" value="921600"/>
        <param name="app_id"        type="int"      value="1024795"/>
        <param name="app_version"   type="int"      value="1"/>
        <param name="enc_key"       type="string"   value="a5f11a418f3dcdc39c4a1efb35a9dd4545e47b56cae9d98c988468277a9305dc"/>
        <param name="sensor_mode" type="bool" value="false"/>
        <param name="align_with_fmu" type="bool" value="false"/>
    </node>

    <node pkg="uart_odom" type="uart_odom" name="uart_odom_client" output="log" respawn="true">
        <!--publishers-->
        <remap from="~out_odom_uav" to="$(arg topic_odom_uav)"/>
        <remap from="~out_odom_cone" to="$(arg topic_odom_cone)"/>
        <!--params-->
        <rosparam command="load" file="$(find uart_odom)/config/uart_odom_config.yaml"/>
        <param name="work_mode" value="real-time"/>
        <param name="js_ctrl_mode" value="feedback"/>
        <param name="serial_port" value="/dev/tty.uwb.slave"/>
    </node>

    <node pkg="n3ctrl" type="n3ctrl_node" name="n3ctrl" output="screen" required="true">
        <!--subscribers-->
		<remap from="~joy" to="/djiros/rc" />
        <remap from="~odom" to="$(arg topic_odom_uav)" />
		<remap from="~imu" to="/djiros/imu" />
		<remap from="~cmd" to="/position_cmd" />
        <!--publishers-->
		<remap from="~desire_pose" to="/desire_pose" />
		<remap from="~traj_start_trigger" to="/traj_start_trigger" />
		<remap from="~ctrl" to="/djiros/ctrl" />
        <!--params-->
        <rosparam command="load" file="$(find n3ctrl)/config/ctrl_param_$(arg uavname).yaml" />
        <param name="work_mode" value="real-time" />
        <param name="js_ctrl_mode" value="feedback" />
        <param name="pub_debug_msgs" type="bool" value="true"/>
	</node>

</launch>
